### ISSUE TITLE
Different implementation of atomic operation abstractions for gcc and clang?

### ISSUE THREAD

**Issue by meyerj** [orocos-toolchain/rtt#259]
Today I stumbled across https://github.com/orocos-toolchain/rtt/commit/94e3053d17352580ae8cac383ec6f7078e13d771, which added a preprocessor directive and switches to different implementations for atomic operations that are essential for lock-free buffer implementations.

I am not aware of the pros and cons of using the builtin intrinsics of gcc/clang compared to the custom platform-dependent implementations. But it would be worth to verify if Clang 3 and newer versions are still affected by the problem which triggered that patch back in 2011. I could not find any background information in the commit history, the [Orocos forums](http://www.orocos.org/orocos/forum) or the [old bug tracker](http://bugs.orocos.org/describecomponents.cgi).

**Comment by snrkiwi** [comment]
I don't think that this is a problem with modern Clang, at least on OS/X or macOS.
